pkg/reconciler/v1alpha1/taskrun/resources/image_exporter_test.go

```diff
@@ -0,0 +1,447 @@
+/*
+Copyright 2018 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/knative/pkg/apis"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	"github.com/tektoncd/pipeline/pkg/logging"
+	"github.com/tektoncd/pipeline/test/names"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestExportingOutputImageResource(t *testing.T) {
+	currentDir, _ := os.Getwd()
+	for _, c := range []struct {
+		desc      string
+		task      *v1alpha1.Task
+		taskRun   *v1alpha1.TaskRun
+		wantSteps []corev1.Container
+	}{{
+		desc: "image resource declared as both input and output",
+		task: &v1alpha1.Task{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "task1",
+				Namespace: "marshmallow",
+			},
+			Spec: v1alpha1.TaskSpec{
+				Inputs: &v1alpha1.Inputs{
+					Resources: []v1alpha1.TaskResource{{
+						Name: "source-image",
+						Type: "image",
+					}},
+				},
+				Outputs: &v1alpha1.Outputs{
+					Resources: []v1alpha1.TaskResource{{
+						Name:            "source-image",
+						Type:            "image",
+						OutputImageDir: currentDir,
+					}},
+				},
+				Steps: []corev1.Container{{
+					Name: "step1",
+				},
+				},
+			},
+		},
+		taskRun: &v1alpha1.TaskRun{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-taskrun-run-output-steps",
+				Namespace: "marshmallow",
+			},
+			Spec: v1alpha1.TaskRunSpec{
+				Inputs: v1alpha1.TaskRunInputs{
+					Resources: []v1alpha1.TaskResourceBinding{{
+						Name: "source-image",
+						ResourceRef: v1alpha1.PipelineResourceRef{
+							Name: "source-image-1",
+						},
+					}},
+				},
+				Outputs: v1alpha1.TaskRunOutputs{
+					Resources: []v1alpha1.TaskResourceBinding{{
+						Name: "source-image",
+						ResourceRef: v1alpha1.PipelineResourceRef{
+							Name: "source-image-1",
+						},
+					}},
+				},
+			},
+		},
+		wantSteps: []corev1.Container{
+			{
+				Name: "step1",
+			},
+			{
+				Name:    "image-digest-exporter-step1-9l9zj",
+				Image:   "override-with-imagedigest-exporter-image:latest",
+				Command: []string{"/ko-app/imagedigestexporter"},
+				Args:    []string{"-images", fmt.Sprintf("[{\"name\":\"source-image-1\",\"type\":\"image\",\"url\":\"gcr.io/some-image-1\",\"digest\":\"\",\"OutputImageDir\":\"%s\"}]", currentDir)},
+			}},
+	}, {
+		desc: "image resource in task with multiple steps",
```

Comment:
i'm trying to understand the difference between the two cases in this test and it seems like:

- both have a resource that is used as an input AND and output
- one has 1 step, the other has 2

I think there might be some more cases we'd want to test, like:
- output resource only
- when the `OutputImageDir` isn't set

This got me wondering if there is a way we could make the `AddOutputImageDigestExporter` function more focused so it would be a bit easier to tell what needed to be tested, and the test could be shorter: what if the function was divided up like this:
- A function that can return all (if any) output image resources from a Task
- A function that sets the `OutputImageDir` as needed
- A function that adds the imageDigestExporterContainer

?

(I don't mind if you want to do that in a follow-up PR instead!)

Reply:
when the `outputImageDir` is not set has a test in the task defaults since it get set there and output resource only don't would make any difference 

for the refactor, makes sense, but I'd prefer if we do it in a follow up PR

Reply:
> for the refactor, makes sense, but I'd prefer if we do it in a follow up PR

kk!